python/pylibvw.cc

```diff
@@ -52,13 +58,83 @@ const size_t pPROB = 6;
 const size_t pMULTICLASSPROBS = 7;
 const size_t pDECISION_SCORES = 8;
 
-
 void dont_delete_me(void*arg) { }
 
-vw_ptr my_initialize(std::string args)
-{ if (args.find_first_of("--no_stdin") == std::string::npos)
+class PyCppBridge : public RED_PYTHON::ExternalBinding {
+  private:
+      py::object* py_reduction_impl;
+      void* base_learner;
+      bool register_finish_learn = false;
+
+  public:
+      int random_num = 0;
+
+      PyCppBridge(py::object* py_reduction_impl) : py_reduction_impl(new py::object(*py_reduction_impl))
+      { this->register_finish_learn = py::extract<bool>(call_py_impl_method("_is_finish_example_implemented"));
+      }
+
+      ~PyCppBridge() { }
+
+      bool ShouldRegisterFinishExample()
+      { return this->register_finish_learn;
+      }
+
+      void SetRandomNumber(int n)
+      { random_num = n;
+      }
+
+      template<typename... Args>
+      py::object call_py_impl_method(char const* method_name, Args&&... args)
+      { try
+        {
+          return this->py_reduction_impl->attr(method_name)(std::forward<Args>(args)...);
+        }
+        catch (...)
+        {
+          // TODO: Properly translate and return Python exception. #2169
+          PyErr_Print();
+          PyErr_Clear();
+          THROW("Exception when calling into python method: " << method_name);
+        }
+      }
+
+      void ActualLearn(example* ec)
+      { this->call_py_impl_method("_learn_convenience", example_ptr(ec, dont_delete_me), py_cpp_bridge_ptr(this, dont_delete_me));
+      }
+
+      void ActualPredict(example* ec)
+      { this->call_py_impl_method("_predict_convenience", example_ptr(ec, dont_delete_me), py_cpp_bridge_ptr(this, dont_delete_me));
+      }
+
+      void ActualFinishExample(example* ec)
+      { this->call_py_impl_method("_finish_example", example_ptr(ec, dont_delete_me));
+      }
+
+      void SetBaseLearner(void* learner)
+      { this->base_learner = learner;
+      }
+
+      void CallBaseLearner(example* ec, bool should_call_learn = true)
```

Comment:
Undecided about this for now, but want to avoid to have as little entrypoints from the python library as possible for now.